mne/io/eeglab/eeglab.py

```diff
@@ -72,32 +104,47 @@ def _get_info(eeg, montage, eog=()):
+
+        # Assume data did not come from an hdf file unless proven so
+        hdf5_flag = False
+
         if (isinstance(eeg.chanlocs, np.ndarray) and not isinstance(
                 eeg.chanlocs[0], io.matlab.mio5_params.mat_struct)):
-            has_pos = all(fld in eeg.chanlocs[0].dtype.names
-                          for fld in pos_fields)
+            try:
+                has_pos = all(fld in eeg.chanlocs[0].dtype.names
+                              for fld in pos_fields)
+            except TypeError:
+                # when stored as hdf, empty eeg.chanlocs is
+                # array([0, 0], dtype=uint64)
+                # which raises a Type Error, but we still need
+                # default chan names
+                ch_names = ["EEG %03d" % ii for ii in range(eeg.nbchan)]
+                hdf5_flag = True
         else:
             has_pos = all(hasattr(eeg.chanlocs[0], fld)
                           for fld in pos_fields)
-        get_pos = has_pos and montage is None
-        pos_ch_names, ch_names, pos = list(), list(), list()
-        kind = 'user_defined'
-        update_ch_names = False
-        for chanloc in eeg.chanlocs:
-            ch_names.append(chanloc.labels)
-            if get_pos:
-                loc_x = _to_loc(chanloc.X)
-                loc_y = _to_loc(chanloc.Y)
-                loc_z = _to_loc(chanloc.Z)
-                locs = np.r_[-loc_y, loc_x, loc_z]
-                if not np.any(np.isnan(locs)):
-                    pos_ch_names.append(chanloc.labels)
-                    pos.append(locs)
-        n_channels_with_pos = len(pos_ch_names)
-        info = create_info(ch_names, eeg.srate, ch_types='eeg')
-        if n_channels_with_pos > 0:
-            selection = np.arange(n_channels_with_pos)
-            montage = Montage(np.array(pos), pos_ch_names, kind, selection)
+        if not hdf5_flag:
+            # If type error thrown by, presumably, hdf5 file, skip the rest
+            # of the if len(eeg.chanlocs) > 0 block
+            get_pos = has_pos and montage is None
+            pos_ch_names, ch_names, pos = list(), list(), list()
+            kind = 'user_defined'
+            update_ch_names = False
+            for chanloc in eeg.chanlocs:
+                ch_names.append(chanloc.labels)
+                if get_pos:
+                    loc_x = _to_loc(chanloc.X)
+                    loc_y = _to_loc(chanloc.Y)
+                    loc_z = _to_loc(chanloc.Z)
+                    locs = np.r_[-loc_y, loc_x, loc_z]
+                    if not np.any(np.isnan(locs)):
+                        pos_ch_names.append(chanloc.labels)
+                        pos.append(locs)
+            n_channels_with_pos = len(pos_ch_names)
+            info = create_info(ch_names, eeg.srate, ch_types='eeg')
+            if n_channels_with_pos > 0:
+                selection = np.arange(n_channels_with_pos)
+                montage = Montage(np.array(pos), pos_ch_names, kind, selection)
```

Comment:
why is the montage part only for non HDF5 file?

Reply:
As I recall, my thinking was that the hdf5_flag got thrown when the hdf5 file didn't store chanlocs (i.e. eeg.chanlocs is empty). So the montage part would be best handled by the `if montage is None` section, since I used the default channel names after the TypeError was thrown and hdf_flag set to True. Was that reasonable?

Reply:
please check as I really don't see why this can be necessary. an old .set file or a new hdf5 contain exactly the same information. So besides the reading to get something that looks like the output of io.loadmat I really don't see why the code should be different.

to see if there is an option to share the maintenance load can you see if https://gitlab.com/obob/pymatreader does a good job for eeglab files?

Reply:
They do contain the same information, it's just an empty eeg.chanlocs was stored as array([0,0], dtype=uint64), which does not have a zero length. So, when you look at the code (from _get_info, lines 72-103)

    if len(eeg.chanlocs) > 0: #This branch is where an empty eeg.chanlocs from hdf5 file wrongly goes
        pos_fields = ['X', 'Y', 'Z']
        if (isinstance(eeg.chanlocs, np.ndarray) and not isinstance(
                eeg.chanlocs[0], io.matlab.mio5_params.mat_struct)):
            has_pos = all(fld in eeg.chanlocs[0].dtype.names
                          for fld in pos_fields)
        else:
            has_pos = all(hasattr(eeg.chanlocs[0], fld)
                          for fld in pos_fields)
        get_pos = has_pos and montage is None
        pos_ch_names, ch_names, pos = list(), list(), list()
        kind = 'user_defined'
        update_ch_names = False
        for chanloc in eeg.chanlocs:
            ch_names.append(chanloc.labels)
            if get_pos:
                loc_x = _to_loc(chanloc.X)
                loc_y = _to_loc(chanloc.Y)
                loc_z = _to_loc(chanloc.Z)
                locs = np.r_[-loc_y, loc_x, loc_z]
                if not np.any(np.isnan(locs)):
                    pos_ch_names.append(chanloc.labels)
                    pos.append(locs)
        n_channels_with_pos = len(pos_ch_names)
        info = create_info(ch_names, eeg.srate, ch_types='eeg')
        if n_channels_with_pos > 0:
            selection = np.arange(n_channels_with_pos)
            montage = Montage(np.array(pos), pos_ch_names, kind, selection)
    elif isinstance(montage, string_types):
        path = op.dirname(montage)
    else:  # if eeg.chanlocs is empty, we still need default chan names
             # This is where empty eeg.chanlocs from hdf5 file should go 
        ch_names = ["EEG %03d" % ii for ii in range(eeg.nbchan)]

I really needed to redirect the code execution to the `else` branch of the if, since it had incorrectly landed in the `if` branch. I used the resulting `TypeError` thrown by the eeg.chanlocs variable, which was mistakenly thought to be not-empty, to execute the code in the `else` block and to create a flag that would let me exit the 'if' branch of the `if-then-else` block.

Reply:
I'll look at `pymatreader` later this week.

Reply:
I was thinking last night that, perhaps, the real answer is in _h5.py to check for whether eeg.chanlocs should really be empty and then, if necessary, make it so. This way the logic in _get_info can return to its original form.